Update transforms on SVG shapes/groups when root element size changed

nt1m opened the pull request:
#### 32128c3f2e5789599a6bcf2eb91c9f8cba54815c
<pre>
Update transforms on SVG shapes/groups when root element size changed
<a href="https://bugs.webkit.org/show_bug.cgi?id=245120">https://bugs.webkit.org/show_bug.cgi?id=245120</a>
&lt;rdar://99854279&gt;

Reviewed by Simon Fraser.

We need to also update transforms when the size of the RenderSVGRoot has changed, since percentage based transforms are relative to the root.

Tests:
- imported/w3c/web-platform-tests/css/css-transforms/transform-percent-010.html
- imported/w3c/web-platform-tests/css/css-transforms/transform-box/view-box-mutation-001.html
- imported/w3c/web-platform-tests/css/css-transforms/transform-box/view-box-mutation-003.html

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::layoutChildren):

Canonical link: <a href="https://commits.webkit.org/254538@main">https://commits.webkit.org/254538@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/592e2c11e614367b9dd91ab600305cfdfceac5f1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89372 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/33930 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/20133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/98702 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/155008 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93380 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32435 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/27939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/81757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93119 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95019 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/25764 "Passed tests") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/76295 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25702 "layout-tests (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80643 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/20133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68689 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30200 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/20133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/29932 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/20133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3185 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33380 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/76295 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/32088 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/20133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
```diff
@@ -267,7 +267,7 @@ void SVGRenderSupport::layoutChildren(RenderElement& start, bool selfNeedsLayout
                     svgText.setNeedsPositioningValuesUpdate();
                 }
-
+                child.setNeedsTransformUpdate();
```
What are the performance implications of always calling this?

nt1m replied:
this is only happening when the RenderSVGRoot changes size.

nikolaszimmermann replied:
Eww, be careful, if the RenderSVGRoot size changes, you are now seting this flag on a all nodes of certain types (e.g. all <rect> elemtns etc.) That is because "element.hasRelativeLengths()" unconditionally returns true here:
```
svg/SVGCircleElement.h:    bool selfHasRelativeLengths() const final { return true; }
svg/SVGEllipseElement.h:    bool selfHasRelativeLengths() const final { return true; }
svg/SVGFilterElement.h:    bool selfHasRelativeLengths() const final { return true; }
svg/SVGForeignObjectElement.h:    bool selfHasRelativeLengths() const final { return true; }
svg/SVGImageElement.h:    bool selfHasRelativeLengths() const final { return true; }
svg/SVGMaskElement.h:    bool selfHasRelativeLengths() const final { return true; }
svg/SVGPatternElement.h:    bool selfHasRelativeLengths() const final { return true; }
svg/SVGRectElement.h:    bool selfHasRelativeLengths() const final { return true; }
```

That was changed in 2014 by Dirk (see 6d66047e8a4c9), see
```
    Since we turned width and height to presentation attributes, the layout optimization
    selfHasRelativeLengths() in the DOM can't be used anymore. selfHasRelativeLengths() was
    intended to solve a problem where we did not layout relatively position/sized elements
    when the parent changes its size. However, as a side effect it did not call layout
    for absolutely positioned/sized elements since the layout does not change. I run
    all performance tests that we have and even wrote a test with hundreds of elements
    that would be affected by this optimization. The differences were inside the sigma
    of a normal test run. (Means I couldn't measure a performance difference.)
```

IMHO the conclusion is wrong though, since width/height are presentation attributes, selfHasRelativeLengths() mechanism should be unified/combined with the possibility to check the CSSValues for percentage units -- because we only ever want to layout() a SVG renderer upon window size changes, if it acually depends on the window size, everything else is an excuse. And I also doubt that this is within one sigma in general -- I could construct a pathological case, where the transform recomputation overhead gets significant (one needs many transformed elements with absolute units, that usually wouldn't need layout, but do because of the selfHasRelativeLengths()==true hack).

I would strongly recommend to measure the impact of the patch and eventually fix the relative length mechanism to consider CSS properties too.

nt1m committed:
Update transforms on SVG shapes/groups when root element size changed
https://bugs.webkit.org/show_bug.cgi?id=245120
<rdar://99854279>

Reviewed by Simon Fraser.

We need to also update transforms when the size of the RenderSVGRoot has changed, since percentage based transforms are relative to the root.

Tests:
- imported/w3c/web-platform-tests/css/css-transforms/transform-percent-010.html
- imported/w3c/web-platform-tests/css/css-transforms/transform-box/view-box-mutation-001.html
- imported/w3c/web-platform-tests/css/css-transforms/transform-box/view-box-mutation-003.html

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::layoutChildren):

Canonical link: https://commits.webkit.org/254538@main
